File: Assets/Scripts/LevelGenerator.js
#pragma strict
#pragma downcast

public var WallCube : GameObject;
public var Monster : GameObject;
public var Player : GameObject;
public var Goal : GameObject;

public var levels : GameObject[];

private var levelObjects : Array;

private var currentLevel : int;
private var totalLevels : int;

function Start () {
	levelObjects = new Array();
	currentLevel = 0;
	totalLevels = levels.Length;
	buildMap();
}

function Update () {
	if (Input.GetKeyDown("r")) restart();
	if (Input.GetKeyDown("n")) startNextLevel();
	if (Input.GetKeyDown("p")) startPreviousLevel();
}

function isWhite(c : Color) {
	return c.r >= 0.9 && c.g >= 0.9 && c.b >= 0.9;
}

function isRed(c : Color) {
	return c.r >= 0.9 && c.g <= 0.1 && c.b <= 0.1;
}

function isGreen(c: Color) {
	return c.r <= 0.1 && c.g >= 0.9 && c.b <= 0.1;
}

function isBlue(c: Color) {
	return c.r <= 0.1 && c.g <= 0.1 && c.b >= 0.9;
}

function isPurple(c: Color) {
	return c.r >= 0.9 && c.g <= 0.1 && c.b >= 0.9;
}

function buildMap() {
	var WIDTH = 16;
	var HEIGHT = 16;
	var SCALE = 32;
	var HALF_SCALE = SCALE / 2;

	var pixels : Array = currentMap().GetPixels();
	
	for (var y = 0; y < HEIGHT; y++) {
		for (var x = 0; x < WIDTH; x++) {
			var i = (y * WIDTH) + x;
			var position = Vector3(x * SCALE + HALF_SCALE, 0, y * SCALE + HALF_SCALE);
			var sPosition = Vector3(273, 0, 528);
			var c : Color = pixels[i];
			if (isWhite(c)) {
				instantiatePrefab(WallCube, position);
			}
			if (isRed(c)) {
				instantiatePrefabScaled(Monster, position, levelParameters().monsterSize);
			}
			if (isGreen(c)) {
				instantiatePrefabScaled(Player, position, levelParameters().playerSize);
			}
			if (isBlue(c)) {
				instantiatePrefab(Goal, position);
			}   
			
			//Secret hackery to unlock special levels
			if (isPurple(c)) {
			instantiatePrefab(Goal, sPosition);
			
				
			}     
		}
	}	
}

function currentMap() {
	return levels[currentLevel].
		GetComponent(LevelController).
		map;
}

function levelParameters() {
	return levels[currentLevel].
		GetComponent(Constants);
}

function restart() {
	destroyLevelObjects();
	buildMap();
}

function startNextLevel() {
	currentLevel = (currentLevel + 1) % totalLevels;
	restart();
}

function startPreviousLevel() {
	currentLevel = (currentLevel == 0 ? totalLevels : currentLevel) - 1;
	restart();
}

function instantiatePrefab(prefab : GameObject, position : Vector3) {
	var instance : GameObject = Instantiate(prefab, position, Quaternion.identity);
	instance.transform.localPosition.y = instance.transform.localScale.y * 0.5;
	addLevelObject(instance);
	return instance;
}

function instantiatePrefabScaled(prefab : GameObject, position : Vector3, size: float) {
	var instance : GameObject = instantiatePrefab(prefab, position);
	instance.transform.localScale = Vector3(size, size, size);
	return instance;
}

// Add a GameObject to be destroyed along with the level.
function addLevelObject(object : GameObject) {
	levelObjects.Add(object);
}

// Destroy level objects, clearing the scene ready for a new level.
function destroyLevelObjects() {
	while (levelObjects.length) {
		Destroy(levelObjects.pop());
	}
}

File: Assets/Scripts/LevelGenerator.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class LevelGenerator : MonoBehaviour
{
    // Start is called before the first frame update
    public GameObject WallCube;
    public GameObject Monster;
    public GameObject Player;
    public GameObject Goal;

    public GameObject[] levels;
    private List<GameObject> levelObjects;

    private int currentLevel;
    private int totalLevels; 



    void Start()
    {
        levelObjects = new List<GameObject>();
        currentLevel = 0;
        totalLevels = levels.Length;
        BuildMap();   
    }

    // Update is called once per frame
    void Update()
    {
        if (Input.GetKeyDown(KeyCode.R)) Restart();
        if (Input.GetKeyDown(KeyCode.N)) StartNextLevel();
        if (Input.GetKeyDown(KeyCode.P)) StartPreviousLevel();

    }

    bool IsWhite(Color c)
    {
        return c.r >= 0.9 && c.g >= 0.9 && c.b >= 0.9;

    }

    bool IsRed(Color c)
    {
        return c.r >= 0.9 && c.g <= 0.1 && c.b <= 0.1;

    }

    bool IsGreen(Color c)
    {
        return c.r <= 0.1 && c.g >= 0.9 && c.b <= 0.1;

    }

    bool IsBlue(Color c)
    {
        return c.r <= 0.1 && c.g <= 0.1 && c.b >= 0.9;

    }

    bool IsPurple(Color c)
    {
        return c.r >= 0.9 && c.g <= 0.1 && c.b >= 0.9;

    }

    void BuildMap()
    {
        var WIDTH = 16;
        var HEIGHT = 16;
        var SCALE = 32;
        var HALF_SCALE = SCALE / 2;

        Color[] pixels = CurrentMap().GetPixels();

        for (int y = 0; y < HEIGHT; y++)
        {
            for(int x = 0; x < WIDTH; x++)
            {
                var i = (y * WIDTH) + x;
                var position = new Vector3(x * SCALE + HALF_SCALE, 0, y * SCALE + HALF_SCALE);
                var sPosition = new Vector3(273, 0, 528);
                Color c = pixels[i];

                if (IsWhite(c))
                {
                    InstantiatePrefab(WallCube, position);
                }
                if (IsRed(c))
                {
                    InstantiatePrefabScaled(Monster, position, LevelParameters().monsterSize);
                }
                if (IsGreen(c))
                {
                    InstantiatePrefabScaled(Player, position, LevelParameters().playerSize);
                }
                if (IsBlue(c))
                {
                    InstantiatePrefab(Goal, position);
                }
                if (IsPurple(c))
                {
                    // Secret hack to unlock special levels
                    InstantiatePrefab(Goal, sPosition);
                }
            }

        }

    }



    public Texture2D CurrentMap()
    {

        return levels[currentLevel].GetComponent<LevelData>().map;
    }

    public Constants LevelParameters()
    {
        return levels[currentLevel].GetComponent<Constants>();

    }


    private void Restart()
    {
        DestroyLevelObjects();
        BuildMap();
    }

    private void StartNextLevel()
    {
        currentLevel = (currentLevel + 1) % totalLevels;
        Restart();

    }

    private void StartPreviousLevel()
    {
        currentLevel = (currentLevel == 0 ? totalLevels : currentLevel) - 1;
        Restart();
    }




    GameObject InstantiatePrefab (GameObject prefab, Vector3 pos)
    {
        var instance = Instantiate(prefab, pos, new Quaternion(), gameObject.transform);
        //  instance.transform.localPosition.y = instance.transform.localScale.y * 0.5;
        AddLevelObject(instance);
        return instance;

    }

    GameObject InstantiatePrefabScaled(GameObject prefab, Vector3 pos, float size)
    {
        GameObject instance = InstantiatePrefab(prefab, pos);
        instance.transform.localScale =  new Vector3(size, size, size);
        return instance;
    }


    void AddLevelObject(GameObject objects)
    {
        levelObjects.Add(objects);

    }

    private void DestroyLevelObjects()
    {
        foreach( GameObject obj in levelObjects)
        {
            Destroy(obj);
        }
        levelObjects.Clear();
    }
}
